src/Sylius/Bundle/AdminBundle/Controller/ImpersonateUserController.php

```diff
@@ -0,0 +1,118 @@
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\Session\Session;
+use Symfony\Component\HttpKernel\Exception\HttpException;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+use Symfony\Component\Translation\TranslatorInterface;
+
+/**
+ * @author Jan Góralski <jan.goralski@lakion.com>
+ */
+class ImpersonateUserController
+{
+    /**
+     * @var UserImpersonatorInterface
+     */
+    protected $impersonator;
+
+    /**
+     * @var AuthorizationCheckerInterface
+     */
+    protected $authorizationChecker;
+
+    /**
+     * @var UserProviderInterface
+     */
+    protected $userProvider;
+
+    /**
+     * @var RouterInterface
+     */
+    protected $router;
+
+    /**
+     * @var TranslatorInterface
+     */
+    protected $translator;
+
+    /**
+     * @var string
+     */
+    protected $authorizationRole;
+
+    /**
+     * @param UserImpersonatorInterface $impersonator
+     * @param AuthorizationCheckerInterface $authorizationChecker
+     * @param UserProviderInterface $userProvider
+     * @param RouterInterface $router
+     * @param TranslatorInterface $translator
+     * @param string $authorizationRole
+     */
+    public function __construct(
+        UserImpersonatorInterface $impersonator,
+        AuthorizationCheckerInterface $authorizationChecker,
+        UserProviderInterface $userProvider,
+        RouterInterface $router,
+        TranslatorInterface $translator,
+        $authorizationRole
+    ) {
+        $this->impersonator = $impersonator;
+        $this->authorizationChecker = $authorizationChecker;
+        $this->userProvider = $userProvider;
+        $this->router = $router;
+        $this->translator = $translator;
+        $this->authorizationRole = $authorizationRole;
+    }
+
+    /**
+     * @param Request $request
+     * @param string $username
+     *
+     * @return Response
+     */
+    public function impersonateAction(Request $request, $username)
+    {
+        if (!$this->authorizationChecker->isGranted($this->authorizationRole)) {
+            throw new HttpException(Response::HTTP_UNAUTHORIZED);
+        }
+
+        $user = $this->userProvider->loadUserByUsername($username);
+        if (null === $user) {
+            throw new HttpException(Response::HTTP_NOT_FOUND);
+        }
+
+        $this->impersonator->impersonate($user);
+
+        $this->addFlash($request, $username);
+
+        return new RedirectResponse($request->headers->get('referer'));
+    }
+
+    /**
+     * @param Request $request
+     * @param string $username
+     */
+    private function addFlash(Request $request, $username)
+    {
+        /** @var Session $session */
+        $session = $request->getSession();
+        $session->getFlashBag()->add('success', $this->translator->trans('sylius.impersonation.success', ['%name%' => $username], 'flashes'));
```

Comment:
I think we should translate the string in the template. You should pass array `['message' => 'sylius.user.impersonate.success', 'parameters' => $username]`.

Btw. discuss translation key of this flash with @GSadee.